drivers/espi/espi_it8xxx2.c

```diff
@@ -0,0 +1,1278 @@
+
+#define IT8XXX2_ESPI_INTERRUPT_ENABLE          BIT(7)
+#define IT8XXX2_ESPI_TO_WUC_ENABLE             BIT(4)
+#define IT8XXX2_ESPI_VW_INTERRUPT_ENABLE       BIT(7)
+#define IT8XXX2_ESPI_INTERRUPT_PUT_PC          BIT(7)
+
+#define IT8XXX2_ESPI_UPSTREAM_INTERRUPT_ENABLE BIT(5)
+#define IT8XXX2_ESPI_UPSTREAM_CHANNEL_DISABLE  BIT(2)
+#define IT8XXX2_ESPI_UPSTREAM_DONE             BIT(1)
+
+#define IT8XXX2_ESPI_PUT_OOB_STATUS            BIT(7)
+#define IT8XXX2_ESPI_PUT_OOB_INTERRUPT_ENABLE  BIT(7)
+
+#define IT8XXX2_ESPI_INPUT_PAD_GATING          BIT(6)
+
+struct espi_it8xxx2_config {
+	uintptr_t base_espi_slave;
+	uintptr_t base_espi_vw;
+	uintptr_t base_espi_queue1;
+	uintptr_t base_espi_queue0;
+	uintptr_t base_ec2i;
+	uintptr_t base_kbc;
+	uintptr_t base_pmc;
+};
+
+struct espi_it8xxx2_data {
+	sys_slist_t callbacks;
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) ((const struct espi_it8xxx2_config *)(dev)->config)
+#define DRV_DATA(dev) ((struct espi_it8xxx2_data *)(dev)->data)
+
+struct vw_channel_t {
+	uint8_t  vw_index;      /* VW index of signal */
+	uint8_t  level_mask;    /* level bit of signal */
+	uint8_t  valid_mask;    /* valid bit of signal */
+};
+
+struct vwidx_isr_t {
+	void (*vwidx_isr)(const struct device *dev, uint8_t update_flag);
+	uint8_t vw_index;
+};
+
+struct espi_isr_t {
+	void (*espi_isr)(const struct device *dev, bool enable);
+	bool asserted_flag;
+};
+
+struct espi_vw_signal_t {
+	enum espi_vwire_signal signal;
+	void (*vw_signal_isr)(const struct device *dev);
+};
+
+/* EC2I bridge and PNPCFG devices */
+static const struct ec2i_t kbc_settings[] = {
+	/* Select logical device 06h(keyboard) */
+	{HOST_INDEX_LDN, LDN_KBC_KEYBOARD},
+	/* Set IRQ=01h for logical device */
+	{HOST_INDEX_IRQNUMX, 0x01},
+	/* Configure IRQTP for KBC. */
+	/*
+	 * Interrupt request type select (IRQTP) for KBC.
+	 * bit 1, 0: IRQ request is buffered and applied to SERIRQ
+	 *        1: IRQ request is inverted before being applied to SERIRQ
+	 * bit 0, 0: Edge triggered mode
+	 *        1: Level triggered mode
+	 *
+	 * This interrupt configuration should the same on both host and EC side
+	 */
+	{HOST_INDEX_IRQTP, 0x02},
+	/* Enable logical device */
+	{HOST_INDEX_LDA, 0x01},
+
+	/* Select logical device 05h(mouse) */
+	{HOST_INDEX_LDN, LDN_KBC_MOUSE},
+	/* Set IRQ=0Ch for logical device */
+	{HOST_INDEX_IRQNUMX, 0x0C},
+	/* Enable logical device */
+	{HOST_INDEX_LDA, 0x01},
+};
+
+static const struct ec2i_t pmc1_settings[] = {
+	/* Select logical device 11h(PM1 ACPI) */
+	{HOST_INDEX_LDN, LDN_PMC1},
+	/* Set IRQ=00h for logical device */
+	{HOST_INDEX_IRQNUMX, 0x00},
+	/* Enable logical device */
+	{HOST_INDEX_LDA, 0x01},
+};
+
+static void ec2i_it8xxx2_wait_status_cleared(const struct device *dev,
+						uint8_t mask)
+{
+	const struct espi_it8xxx2_config *const config = DRV_CONFIG(dev);
+	struct ec2i_regs *const ec2i = (struct ec2i_regs *)config->base_ec2i;
+
+	while (ec2i->IBCTL & mask) {
+		;
+	}
```

Comment:
Any way this will never get set?

Reply:
The Vcc is configured as high so we can always get the status we want.

Reply:
In the cros-ec driver, this function delays 15 us and then reads the IBCTL register once - with no polling.  What's the correct approach?

Reply:
The polling is correct approach.
EC2I write operation sequence is recommend as folllowing in datasheet:
1. Set CSAE bit in IBCTL register.
2. Make sure that both CRIB and CWIB bits in IBCTL register are cleared.
3. Setting its enable bit in IBMAE register for access module, only one module can enable at a time.
4. Assign the offset of the register in the device in IHIOA register.
5. Write the data to IHD register, which begins a write transaction.
6. Read the CWIB bit in IBCTL until it returns 0, which represents that a write transaction has been finished.